### ISSUE TITLE
WebViewGoogle causes crashes on Nougat

### ISSUE THREAD

**Issue by jamuir** [opengapps/aosp_build#89]
Including WebViewGoogle in N5X Nougat builds causes many apps to crash.

Google Setup wizard crashes and other apps that require webview crash.

There is logcat output below.  Note the following line:

```
09-28 09:56:31.714 E/WebViewFactory( 4825): Caused by: java.lang.UnsatisfiedLinkError: dlopen failed: "/system/app/Chrome/Chrome.apk!/lib/armeabi-v7a/libmonochrome.so" is 32-bit instead of 64-bit
```

```
09-28 09:56:31.714 E/WebViewFactory( 4825): u0 : error instantiating provider
09-28 09:56:31.714 E/WebViewFactory( 4825): java.lang.reflect.InvocationTargetException
09-28 09:56:31.714 E/WebViewFactory( 4825):     at java.lang.reflect.Constructor.newInstance0(Native Method)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at java.lang.reflect.Constructor.newInstance(Constructor.java:430)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at android.webkit.WebViewFactory.getProvider(WebViewFactory.java:198)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at android.webkit.CookieManager.getInstance(CookieManager.java:39)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at com.google.android.apps.gsa.search.core.d.Ud(Cookies.java:1256)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at com.google.android.apps.gsa.search.core.d.Ue(Cookies.java:506)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at com.google.android.apps.gsa.extradex.backgroundtasks.a.run(RefreshSearchDomainAndCookiesTask.java:4286)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at com.google.android.apps.gsa.search.core.tasks.b.a$3.run(LegacySearchBackgroundTaskFactory.java:614)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at java.util.concurrent.FutureTask.run(FutureTask.java:237)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at java.util.concurrent.FutureTask.run(FutureTask.java:237)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at com.google.android.apps.gsa.shared.util.concurrent.a.ag.run(TaskRunnerFutureTask.java:122)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at com.google.android.apps.gsa.shared.util.concurrent.a.ar.run(ThrottlingExecutor.java:139)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at com.google.android.apps.gsa.shared.util.concurrent.a.ar.run(ThrottlingExecutor.java:139)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at java.lang.Thread.run(Thread.java:761)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at com.google.android.apps.gsa.shared.util.concurrent.a.aa$1.run(GsaThreadFactory.java:85)
09-28 09:56:31.714 E/WebViewFactory( 4825): Caused by: java.lang.RuntimeException: Cannot load WebView
09-28 09:56:31.714 E/WebViewFactory( 4825):     at org.chromium.android_webview.AwBrowserProcess.loadLibrary(AwBrowserProcess.java:54)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at com.android.webview.chromium.WebViewChromiumFactoryProvider.initialize(WebViewChromiumFactoryProvider.java:147)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at com.android.webview.chromium.WebViewChromiumFactoryProvider.<init>(WebViewChromiumFactoryProvider.java:115)
09-28 09:56:31.714 E/WebViewFactory( 4825):     ... 19 more
09-28 09:56:31.714 E/WebViewFactory( 4825): Caused by: org.chromium.base.library_loader.ProcessInitException
09-28 09:56:31.714 E/WebViewFactory( 4825):     at org.chromium.base.library_loader.LibraryLoader.loadAlreadyLocked(LibraryLoader.java:308)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at org.chromium.base.library_loader.LibraryLoader.loadNow(LibraryLoader.java:168)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at org.chromium.android_webview.AwBrowserProcess.loadLibrary(AwBrowserProcess.java:48)
09-28 09:56:31.714 E/WebViewFactory( 4825):     ... 21 more
09-28 09:56:31.714 E/WebViewFactory( 4825): Caused by: java.lang.UnsatisfiedLinkError: dlopen failed: "/system/app/Chrome/Chrome.apk!/lib/armeabi-v7a/libmonochrome.so" is 32-bit instead of 64-bit
09-28 09:56:31.714 E/WebViewFactory( 4825):     at java.lang.Runtime.loadLibrary0(Runtime.java:977)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at java.lang.System.loadLibrary(System.java:1530)
09-28 09:56:31.714 E/WebViewFactory( 4825):     at org.chromium.base.library_loader.LibraryLoader.loadAlreadyLocked(LibraryLoader.java:294)
09-28 09:56:31.714 E/WebViewFactory( 4825):     ... 23 more
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825): u0 : Background task refresh_search_domain_then_send_gsa_home_request failed
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825): android.util.AndroidRuntimeException: java.lang.reflect.InvocationTargetException
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at android.webkit.WebViewFactory.getProvider(WebViewFactory.java:204)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at android.webkit.CookieManager.getInstance(CookieManager.java:39)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at com.google.android.apps.gsa.search.core.d.Ud(Cookies.java:1256)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at com.google.android.apps.gsa.search.core.d.Ue(Cookies.java:506)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at com.google.android.apps.gsa.extradex.backgroundtasks.a.run(RefreshSearchDomainAndCookiesTask.java:4286)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at com.google.android.apps.gsa.search.core.tasks.b.a$3.run(LegacySearchBackgroundTaskFactory.java:614)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at java.util.concurrent.FutureTask.run(FutureTask.java:237)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at java.util.concurrent.FutureTask.run(FutureTask.java:237)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at com.google.android.apps.gsa.shared.util.concurrent.a.ag.run(TaskRunnerFutureTask.java:122)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at com.google.android.apps.gsa.shared.util.concurrent.a.ar.run(ThrottlingExecutor.java:139)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at com.google.android.apps.gsa.shared.util.concurrent.a.ar.run(ThrottlingExecutor.java:139)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at java.lang.Thread.run(Thread.java:761)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at com.google.android.apps.gsa.shared.util.concurrent.a.aa$1.run(GsaThreadFactory.java:85)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825): Caused by: java.lang.reflect.InvocationTargetException
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at java.lang.reflect.Constructor.newInstance0(Native Method)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at java.lang.reflect.Constructor.newInstance(Constructor.java:430)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at android.webkit.WebViewFactory.getProvider(WebViewFactory.java:198)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    ... 16 more
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825): Caused by: java.lang.RuntimeException: Cannot load WebView
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at org.chromium.android_webview.AwBrowserProcess.loadLibrary(AwBrowserProcess.java:54)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at com.android.webview.chromium.WebViewChromiumFactoryProvider.initialize(WebViewChromiumFactoryProvider.java:147)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at com.android.webview.chromium.WebViewChromiumFactoryProvider.<init>(WebViewChromiumFactoryProvider.java:115)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    ... 19 more
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825): Caused by: org.chromium.base.library_loader.ProcessInitException
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at org.chromium.base.library_loader.LibraryLoader.loadAlreadyLocked(LibraryLoader.java:308)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at org.chromium.base.library_loader.LibraryLoader.loadNow(LibraryLoader.java:168)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at org.chromium.android_webview.AwBrowserProcess.loadLibrary(AwBrowserProcess.java:48)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    ... 21 more
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825): Caused by: java.lang.UnsatisfiedLinkError: dlopen failed: "/system/app/Chrome/Chrome.apk!/lib/armeabi-v7a/libmonochrome.so" is 32-bit instead of 64-bit
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at java.lang.Runtime.loadLibrary0(Runtime.java:977)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at java.lang.System.loadLibrary(System.java:1530)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    at org.chromium.base.library_loader.LibraryLoader.loadAlreadyLocked(LibraryLoader.java:294)
09-28 09:56:31.718 E/VelvetBackgroundTasksIm( 4825):    ... 23 more
```

There are some comments/references in https://github.com/opengapps/aosp_build/issues/78 that may be relevant.

**Comment by jamuir** [comment]
Some more info:

- I am using the "stock" gapps configuration.
- I've hit this defect in only *some* of my builds, so its manifestation may depend on when the GoogleWebView recipe gets executed during the build.
- the conversation in the following PR is relevant:  https://github.com/opengapps/aosp_build/pull/76

**Comment by jamuir** [comment]
I think the issue is just that the wrong Chrome apk is being installed:

```
bullhead:/ $ md5sum /system/app/Chrome/Chrome.apk                                                                                                                                                                                           
9a3e3b5b75f5c39c9c3c41fb3b9094f7  /system/app/Chrome/Chrome.apk
```

```
vendor/opengapps/sources$ find . -type f | grep -i chrome | xargs md5sum
e73cf58ba38bed8b2ede8aa98987981a  ./arm/app/com.android.chrome/16/nodpi/278512400.apk
9a3e3b5b75f5c39c9c3c41fb3b9094f7  ./arm/app/com.android.chrome/24/nodpi/278512402.apk
c5647153ce5fff5bfd6e8241d88a1d0f  ./arm/app/com.android.chrome/21/nodpi/278512401.apk
825775baeebe5a3ba1a0725ee36d770f  ./arm64/app/com.android.chrome/24/nodpi/.gitignore
d404d8c3f6ea0ea87cff7b1c37995941  ./arm64/app/com.android.chrome/24/nodpi/278512452.apk
d09306715e9c5014d0060637808d379f  ./arm64/app/com.android.chrome/24/nodpi/278512452.apk.lz
```
so you can see that the "arm" Chrome.apk was installed rather than the "arm64" version.

**Comment by aunali1** [comment]
Happening for me aswell.

**Comment by aunali1** [comment]
`09-28 18:30:36.106  3749  3749 W System  : ClassLoader referenced unknown path: /system/app/Chrome/lib/arm
09-28 18:30:36.109  3749  3749 I WebViewFactory: Loading com.android.chrome version 53.0.2785.124 (code 278512402)
09-28 18:30:36.110  3749  3749 E webviewchromiumloader: Failed to open relro file /data/misc/shared_relro/libwebviewchromium64.relro: No such file or directory
09-28 18:30:36.110  3749  3749 W WebViewFactory: failed to load with relro file, proceeding without
09-28 18:30:36.118  3749  3749 E WebViewFactory: error instantiating provider
09-28 18:30:36.118  3749  3749 E WebViewFactory: java.lang.reflect.InvocationTargetException`

**Comment by aunali1** [comment]
Decompressing the .lz file causes it to work for me.

**Comment by jamuir** [comment]
if do

```
$ lunzip vendor/opengapps/sources/arm64/app/com.android.chrome/24/nodpi/278512452.apk.lz
```
and rebuild your device image, then the correct apk will be installed and the error goes away.

**Comment by jamuir** [comment]
I opened a PR that fixes this:  https://github.com/opengapps/aosp_build/pull/90

**Comment by Blystad** [comment]
Should be fixed now that #90 is merged.
